cpp/include/raft/lap/lap.cuh

```diff
@@ -0,0 +1,280 @@
+ *
+ *      Article reference:
+ *          Date, Ketan, and Rakesh Nagi. "GPU-accelerated Hungarian algorithms
+ *          for the Linear Assignment Problem." Parallel Computing 57 (2016): 52-72.
+ *
+ */
+#pragma once
+
+#include <raft/handle.hpp>
+
+#include "d_structs.h"
+#include "lap_functions.cuh"
+
+namespace raft {
+namespace lap {
+
+template <typename vertex_t, typename weight_t>
+class LinearAssignmentProblem {
+  vertex_t size_;
+  vertex_t batchsize_;
+
+  weight_t const *d_costs_;
+
+  Vertices<vertex_t, weight_t> d_vertices_dev;
+  VertexData<vertex_t> d_row_data_dev, d_col_data_dev;
+
+  raft::handle_t const &handle_;
+  raft::mr::device::buffer<int> row_covers_v;
+  raft::mr::device::buffer<int> col_covers_v;
+  raft::mr::device::buffer<weight_t> row_duals_v;
+  raft::mr::device::buffer<weight_t> col_duals_v;
+  raft::mr::device::buffer<weight_t> col_slacks_v;
+  raft::mr::device::buffer<int> row_is_visited_v;
+  raft::mr::device::buffer<int> col_is_visited_v;
+  raft::mr::device::buffer<vertex_t> row_parents_v;
+  raft::mr::device::buffer<vertex_t> col_parents_v;
+  raft::mr::device::buffer<vertex_t> row_children_v;
+  raft::mr::device::buffer<vertex_t> col_children_v;
+  raft::mr::device::buffer<weight_t> obj_val_primal_v;
+  raft::mr::device::buffer<weight_t> obj_val_dual_v;
+
+ public:
+  LinearAssignmentProblem(raft::handle_t const &handle, vertex_t size,
+                          vertex_t batchsize)
+    : handle_(handle),
+      size_(size),
+      batchsize_(batchsize),
+      d_costs_(nullptr),
+      row_covers_v(handle_.get_device_allocator(), handle_.get_stream(), 0),
+      col_covers_v(handle_.get_device_allocator(), handle_.get_stream(), 0),
+      row_duals_v(handle_.get_device_allocator(), handle_.get_stream(), 0),
+      col_duals_v(handle_.get_device_allocator(), handle_.get_stream(), 0),
+      col_slacks_v(handle_.get_device_allocator(), handle_.get_stream(), 0),
+      row_is_visited_v(handle_.get_device_allocator(), handle_.get_stream(), 0),
+      col_is_visited_v(handle_.get_device_allocator(), handle_.get_stream(), 0),
+      row_parents_v(handle_.get_device_allocator(), handle_.get_stream(), 0),
+      col_parents_v(handle_.get_device_allocator(), handle_.get_stream(), 0),
+      row_children_v(handle_.get_device_allocator(), handle_.get_stream(), 0),
+      col_children_v(handle_.get_device_allocator(), handle_.get_stream(), 0),
+      obj_val_primal_v(handle_.get_device_allocator(), handle_.get_stream(), 0),
+      obj_val_dual_v(handle_.get_device_allocator(), handle_.get_stream(), 0) {}
+
+  // Executes Hungarian algorithm on the input cost matrix.
+  void solve(weight_t const *d_cost_matrix, vertex_t *d_row_assignment,
+             vertex_t *d_col_assignment) {
+    initializeDevice();
+
+    d_vertices_dev.row_assignments = d_row_assignment;
+    d_vertices_dev.col_assignments = d_col_assignment;
+
+    d_costs_ = d_cost_matrix;
+
+    int step = 0;
+
+    while (step != 100) {
```

Comment:
this seems ripe for an endless loop situation.  seems like there should be a default case statement to catch issues

Reply:
The algorithm is guaranteed to converge and eventually return 100 unless there's a bug in implementation.  The return values are all constants and never computed:
* hungarianStep0 always returns 1
* hungarianStep1 can only return a 5 or a 6
* hungarianStep2 can only return a 3 or a 6
* hungarianStep3 can only return a 4 or a 5
* hungarianStep4 always returns 2
* hungarianStep5 always returns 3
* hungarianStep6 always returns 100